### ISSUE TITLE
Add possibility to run js-beautify as check with fail on change

### ISSUE THREAD

**Issue by adrianschmidt** [beautify-web/js-beautify#741]
In order to run js-beautify as a non-intrusive pre-commit hook in git (or any other place where you wish to check that the code is beautified, but not actually change it), it would help immensely to have an option not to actually save the beautified version of a file anywhere, but to just have js-beautify fail with a non-zero exit code and some reasonable message.

**Comment by bitwiseman** [comment]
Add a `check-only` or `dry-run` option.    Sure, worth doing. 

Until this gets added, you can do accomplish this (I'm sure you already have) by outputting to separate folder and running a diff.

**Comment by adrianschmidt** [comment]
Currently I'm reading the file and the output into a string and checking them for equality, but the folder-diff way is something I'll definitely try. Then I could have the hook output what needs to be changed too. Thanks for the input, and for adding my suggestion to a milestone!

**Comment by bitwiseman** [comment]
You might also try just grepping the output.  Any file that doesn't include " - unchanged" after it is a failure of your test.  
https://github.com/beautify-web/js-beautify/blob/master/js/lib/cli.js#L324

**Comment by adrianschmidt** [comment]
Yup, I found that in the source already. Only problem is that it requires me to write the changes to a file, or js-beautify won't output that. So if I'm going to write to files, I could just as well do the diff and output that, so the user can see what needs to be fixed, in case they don't want to run js-beautify (for whatever reason, not all my team members are quite as happy running automatic processing on their code as I am).

**Comment by ilovett** [comment]
Would it be possible or worth writing files to a tmp directory and then deleting?

**Comment by mnebuerquo** [comment]
I initially wanted to make js-beautify change the code being committed in a git pre-commit hook, but it would have to stage the modified files after beautifying. This is a problem for committing partial changes, since the user could stage a file, make more changes and then try to commit. Only the staged hunks should be committed, and other js-beautify changes should not be staged at that time. There isn't an easy way to make the pre-commit hook smart about what it stages and what it does not.

A diff option is an excellent solution to the problem because then the user can see what has failed and can run the beautify command (or manually fix it) and stage whatever needs to be committed before retrying the commit.

The esformatter project implements this feature (called with --diff), though it won't also do my html and css files:
https://github.com/millermedeiros/esformatter

**Comment by jtomaszewski** [comment]
I implemented it myself with a simple bash file:

```
#!/bin/bash

APPDIR=$(realpath $(dirname $(dirname $0)))
TMPDIR=$(mktemp -d)
cd $APPDIR

for line in $*; do
  filepath=$(realpath $line)
  file=${filepath/$APPDIR\//}
  mkdir -p $TMPDIR/$(dirname $file)
  cp $APPDIR/$file $TMPDIR/$file
done

message=$(find $TMPDIR -type f | xargs node_modules/.bin/html-beautify -r | grep -v unchanged || true)
if [[ $message ]]; then
  echo "$message" | sed "s#$TMPDIR##g" | sed "s#\.\.\/##g"
  rm -rf $TMPDIR
  exit 1
fi

rm -rf $TMPDIR
```

Save it f.e. as `bin/html-beautify-check` and then run like `bin/html-beautify-check app1.html app2.html app3.html` or `find . -type f | grep .html | xargs bin/html-beautify-check` .

**Comment by stoooops** [comment]
Is this project dead? Seems like a basic feature left unimplemented for five years.

**Comment by mnebuerquo** [comment]
I switched to prettier after I left the comment above, and I never looked back. I don't know if this project is still maintained.

**Comment by adrianschmidt** [comment]
> I switched to prettier after I left the comment above, and I never looked back. I don't know if this project is still maintained.

Yeah, our project also uses prettier, together with eslint.

**Comment by bitwiseman** [comment]
@stoooops 
Would you like to contribute this feature?  Just waiting for someone to do the work.  Feel free.

**Comment by romainnorberg** [comment]
something like that in `package.json`:

```
  "scripts": {
    "test-jslint": "if js-beautify ./assets/js/* -r | grep unchanged; then exit 0; else exit 1; fi"
  }
```

**Comment by kgraefe** [comment]
> something like that in `package.json`:
> 
> ```
>   "scripts": {
>     "test-jslint": "if js-beautify ./assets/js/* -r | grep unchanged; then exit 0; else exit 1; fi"
>   }
> ```

sorry for necromancing. I've been in the same situation today in my CI. The proposed solution checks if *all* files are changed. If only one file is unchanged it returns `0` (success), regardless of the other files.

Luckily `diff` returns `1` if there are changes, so what I do in my CI is:
```
git ls-files '*.js' | xargs -n 1 --verbose bash -c 'diff -ru "$0" <(js-beautify "$0")'
```

(`diff` can be replaced by `cmp` if you're not interested in the differences)
